xwiki-platform-core/xwiki-platform-web/src/main/webapp/templates/adduorg.vm

```diff
@@ -17,22 +17,49 @@
 ## Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 ## 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 ## ---------------------------------------------------------------------------
+##
+## Checks if this is valid to add the addedGroup to the currentGroup.
+## This verification is based on three criteria:
+## - the current and added groups are not the same group
+## - the current group does not already have the added group in its sub-groups
+## - this current group is not sub-group of the added group
+##
+## The boolean result is set on the $accept variable.
+##
+#macro (checkGroups $currentGroup $addedGroup)
+  #set ($members = $services.user.group.getMembers($currentGroup, true))
```

Comment:
This recursive method can be very expensive, I'm not sure using a velocity macro for this is the right approach: IMO this kind of check should rely on some cache mechanisms to avoid computing back everything when it's possible. And it needs to have really good coverage, so I'd provide the check as a Java API.

Reply:
Not sure what you mean, I don't see any Velocity recursion here or maybe I missed it. `getMembers` is what is taking care of the recursion and it's a Java API which rely on caching. Sure we could add a `isMemberOf` helper but it's not going to change much in term of performance, but making the Velocity code more readable is good too.

Reply:
> getMembers is what is taking care of the recursion and it's a Java API which rely on caching.

Was talking about this recursion not about a velocity recursion. I remembered it being expensive but maybe I'm wrong there. In general this `checkGroups` sounds more like something we'd want at Java level than at velocity level: we could need it when performing an add member at API level too for example.

Reply:
> I remembered it being expensive but maybe I'm wrong there.

It could of course be expensive by definition if the member cache is totally empty and the entity is member of a lot of groups but there is not much else we can do, it should not be very noticeable in practice. We could of course have a helper in the API for checkGroups but with a much better name then :)

Reply:
I'm considering moving the logic to a method called `GroupScriptService#canAddNewChild(DocumentReference parent, DocumentReference child)`, and of course taking advantage of this code update to take into account the other remarks. WDYT?

Reply:
Sounds good (not sure it's the best name but I'm not against it except that the naming everywhere is "member" and not "child"). Note that you should keep the check related to empty string on Velocity side since it does not make much sense after the conversion to DocumentReference.

Reply:
The logic is now moved to a dedicated method in the groups script service (see https://github.com/xwiki/xwiki-platform/pull/1374/files#diff-8dc9d53466040f00f533fc828198b1dfR154)